Prevent Lock Blocking by Adding Timeout to `JournalStorage`

sawa3030 opened the pull request:
## Motivation
- Currently, if a process is terminated without releasing the lock, the next process fails to resume. This change introduces a mechanism to release the existing lock if acquiring it takes too long.
- ref: [JournalStorage with JournalFileSymlinkLock may fail to resume #5921](https://github.com/optuna/optuna/issues/5921)

## Description of the changes
- Introduced a grace period for acquiring locks in `JournalFileSymlinkLock` and `JournalFileOpenLock`.
- If the lock cannot be acquired within the grace period, it is considered stale and released to prevent blocking new processes.

y0z commented:
@not522 Could you review this PR?

not522 reviewed (commented):
Thank you for your PR! Could you check the following points?
- How about using `time.monotonic()`? This differs from `datetime.now()` in that it is not affected by system clock updates.
- How about checking the last modified time of the lock? As the number of processes increases, it may continue to fail to acquire a lock.

sawa3030 commented:
@not522 Thank you for your suggestions. There are some changes I made.
- I updated the logic to check the last modified time. The previous approach had issues when the grace period was too short, but with your suggested logic, it is now more robust regardless of the grace period length.
- I have confirmed that this PR successfully resolves the issue: [JournalStorage with JournalFileSymlinkLock may fail to resume #5921](https://github.com/optuna/optuna/issues/5921). After an unexpected process termination, the optimization process can now successfully resume even if the lock file still exists. It successfully resumes the optimization process whether a new process starts before or after the grace period has elapsed.
- I added tests to ensure that a customized grace period functions correctly without issues.

And here are some concerns
- As far as I know, `time.monotonic()` cannot be used for checking the modified time of a file. Do you think it would be better to keep the previous logic, as shown in the code below? Or would that make the implementation unnecessarily complex?
```
if min(datetime.datetime.now() - mtime, time.monotonic() - start_time) > datetime.timedelta(seconds=self.grace_period):
    self.release()
```
- [The documentation for st_mtime](https://docs.python.org/ja/3.13/library/os.html#os.stat_result.st_birthtime_ns) states that its resolution depends on the file system, and some file systems have a low resolution (e.g., 2-seconds). If the grace period is set too short, this could lead to unexpected behavior. I’ve added a warning if the grace period is set to less than 3 seconds, but I’m unsure whether this warning is necessary or if it just adds unnecessary complexity.

c-bata commented:
@y0z Could you review this PR?

not522 commented:
In my understanding, comparing `st_mtime` and the local time of the machine is unreliable (for example, NFS records the time of the server, not the machine where the process is running). But, we can trust whether it has been modified or not. So, how about the following logic?

```python
start_time = time.monotonic()
mtime = datetime.datetime.fromtimestamp(os.stat(self._lock_file).st_mtime)
while True:
    ...
    current_mtime = datetime.datetime.fromtimestamp(os.stat(self._lock_file).st_mtime)
    if current_mtime != mtime:
        start_time = time.monotonic()
        mtime = current_mtime
        continue
    if time.monotonic() - start_time > self.grace_period:
        self.release()
```

sawa3030 commented:
Thank you for your suggestion! I've updated the logic accordingly.

sawa3030 commented:
@not522 @y0z I’ve made some updates. Could you review the PR when you have time? Thank you!

not522 commented on tests/storages_tests/journal_tests/test_journal.py:
```diff
@@ -28,33 +28,48 @@
                 raise Exception("Must not reach here")
             return optuna.storages.journal.JournalFileBackend(self.tempfile.name, lock)
         elif self.storage_type.startswith("redis"):
+            if self.grace_period is not None:
+                raise RuntimeError("Grace period is not supported for Redis storage.")
```
How about just asserting `self.grace_period is None`?

not522 reviewed (commented):
Thank you for your update! Could you check my suggestions?

sawa3030 commented:
@not522 Thank you for your advice! I've made some updates based on your feedback.

not522 commented on optuna/storages/journal/_file.py:
```diff
@@ -128,26 +129,43 @@ class JournalFileSymlinkLock(BaseJournalFileLock):
                 return True
             except OSError as err:
                 if err.errno == errno.EEXIST:
+                    try:
```
I think that it is just for `self.release`, isn't it?

sawa3030 replied:
I overlooked that release() could also raise an error. To answer your quesetion, in my understanding, os.stat() can generate an [OSError](https://docs.python.org/ja/3.13/library/os.html#os.error) if the lock file is removed by another process and no longer exists.

not522 replied:
I see. How about clarifying the target of the try as follows:

```python
try:
    current_mtime = os.stat(self._lock_file).st_mtime
except OSError:
    continue

sawa3030 replied:
I wasn’t aware that this code works without encountering a scope error. Thank you for your step by step suggestion🙇
```
try:
    current_mtime = os.stat(self._lock_file).st_mtime
except OSError:
    continue
if current_mtime != mtime:
    mtime = current_mtime
    last_update_time = time.monotonic()
```

y0z commented on optuna/storages/journal/_file.py:
```diff
@@ -194,13 +228,32 @@ def acquire(self) -> bool:
                 return True
             except OSError as err:
                 if err.errno == errno.EEXIST:
+                    try:
+                        current_mtime = os.stat(self._lock_file).st_mtime
+                    except OSError:
+                        continue
+                    if current_mtime != mtime:
+                        mtime = current_mtime
+                        last_update_time = time.monotonic()
```
I am confused about why we need these three variables: `last_update_time`, `mtime`, and `current_mtime`. At first glance, it seems that `last_update_time = current_mtime` since `current_mtime != mtime` means the lock file has been modified. Could you please provide additional explanation as to the difference between their roles?

sawa3030 replied:
`last_update_time` is based on monotonic clock time, whereas `current_mtime` and `m_time` are Epoch timestamps. I chose not to compare `mtime` directly with grace_period to avoid inconsistencies.

sawa3030 replied:
I feel like the name `last_update_time` is a little confusing, but I couldn't come up with a better variable name.

sawa3030 replied:
I change the variable name from `last_update_time` to `last_update_monotonic_time`!

y0z commented on optuna/storages/journal/_file.py:
```diff
@@ -128,26 +129,52 @@ class JournalFileSymlinkLock(BaseJournalFileLock):
                 return True
             except OSError as err:
                 if err.errno == errno.EEXIST:
+                    try:
+                        current_mtime = os.stat(self._lock_file).st_mtime
+                    except OSError:
+                        continue
+                    if current_mtime != mtime:
+                        mtime = current_mtime
+                        last_update_time = time.monotonic()
```
Ditto: https://github.com/optuna/optuna/pull/5971/files#r1973051800

y0z commented on optuna/storages/journal/_file.py:
```diff
@@ -128,26 +129,52 @@ class JournalFileSymlinkLock(BaseJournalFileLock):
     """
 
-    def __init__(self, filepath: str) -> None:
+    def __init__(self, filepath: str, grace_period: int = 30) -> None:
```
What about allowing `grace_period = None` (i.e., unlimited)? Forcing the release of long-held locks seems like a good practice in many cases, but I'm not sure it's always OK.

y0z commented on optuna/storages/journal/_file.py:
```diff
@@ -128,26 +129,52 @@ class JournalFileSymlinkLock(BaseJournalFileLock):
+                            "for an extended period. Forcibly releasing the lock file."
+                        )
+                        try:
+                            self.release()
```
Would it be a good idea to reset `sleep_secs` to `0.001` after calling `release()`? Another thread may acquire the lock between your `release()` and `os.symlink()`.

y0z commented on optuna/storages/journal/_file.py:
```diff
@@ -181,10 +208,17 @@ class JournalFileOpenLock(BaseJournalFileLock):
     """
 
-    def __init__(self, filepath: str) -> None:
+    def __init__(self, filepath: str, grace_period: int = 30) -> None:
```
Ditto: https://github.com/optuna/optuna/pull/5971/files#r1973119843

sawa3030 commented:
@y0z I made the update. Thank you for your feedback!

sawa3030 commented:
@y0z @not522 I think all updates have been applied. Thank you for your suggestions! Please let me know if there's anything else that needs to be addressed.

not522 commented on optuna/storages/journal/_file.py:
```diff
@@ -128,26 +129,55 @@ class JournalFileSymlinkLock(BaseJournalFileLock):
+                            try:
+                                self.release()
+                                sleep_secs = 0.001
+                            except OSError:
```
Maybe it should be `RuntimeError`.

```suggestion
                            except RuntimeError:
```

not522 reviewed (commented):
Thank you. It's almost LGTM. Could you checking the following points?

- The error type
- Adding test for the invalid argument (i.e., zero and negative integer)

I think the warning message can be improved, but I think we can address in a separate PR.

y0z reviewed (approved):
LGTM

not522 commented on tests/storages_tests/journal_tests/test_journal.py:
```diff
@@ -28,33 +28,43 @@
-    "redis_default",
-    "redis_with_use_cluster",
+LOG_STORAGE_WITH_PARAMETER = [
+    ("file_with_open_lock", 3),
+    ("file_with_open_lock", 30),
+    ("file_with_open_lock", None),
+    ("file_with_link_lock", 3),
+    ("file_with_link_lock", 30),
+    ("file_with_link_lock", None),
```
I think `grace_period=3` is redundant.

```suggestion
    ("file_with_open_lock", 30),
    ("file_with_open_lock", None),
    ("file_with_link_lock", 30),
    ("file_with_link_lock", None),
```

not522 commented on tests/storages_tests/journal_tests/test_journal.py:
```diff
@@ -244,3 +258,18 @@ def test_raise_error_for_deprecated_class_import_from_journal() -> None:
         journal.BaseJournalLogStorage  # type: ignore[attr-defined]
+
+
+@pytest.mark.parametrize(
+    "log_storage_type,grace_period",
+    [
+        ("file_with_open_lock", 0),
+        ("file_with_open_lock", -1),
+        ("file_with_link_lock", 0),
+        ("file_with_link_lock", -1),
+    ],
+)
```
It can be written more simply.

```suggestion
@pytest.mark.parametrize("log_storage_type", ("file_with_open_lock", "file_with_link_lock"))
@pytest.mark.parametrize("grace_period", (0, -1))
```

sawa3030 commented:
@not522 Thank you for letting me know that. I updated the code as you suggested.

not522 reviewed (approved):
LGTM!
